### PR TITLE
[AMDGPU] Fix MC/Disassembler/AMDGPU/decode-err.txt.

### PR DESCRIPTION
**Pull request by kosarev**
It fails downstream now that
https://github.com/llvm/llvm-project/pull/95237 removed flushing the output stream on printing every instruction.

### CODE REVIEW

File: llvm/test/MC/Disassembler/AMDGPU/decode-err.txt [hunk 1]
```diff
@@ -39,11 +41,11 @@
...
-# GFX12: [[@LINE+1]]:1: warning: invalid instruction encoding
+# GFX12-ERR: [[@LINE+1]]:1: warning: invalid instruction encoding
 0x00,0x00,0x56,0xd8,0x00,0x01,0x00,0x00
 
 # this is image_msaa_load where samp field for gfx12 VSAMPLE is not all zeros
-# GFX12: [[@LINE+1]]:1: warning: invalid instruction encoding
+# GFX12-ERR: [[@LINE+1]]:1: warning: invalid instruction encoding
 0x06,0x00,0x46,0xe4,0x01,0x10,0x80,0x00,0x05,0x06,0x07,0x00
 
 # This is ds_read_b32 with gds bit which is not valid on gfx90a.
```
Comment (kosarev) [inline llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:51]: @jayfoad We don't seem to check `GFX90A` prefixes in this test.

File: llvm/test/MC/Disassembler/AMDGPU/decode-err.txt [hunk 2]
```diff
@@ -1,17 +1,19 @@
-# RUN: llvm-mc -triple=amdgcn -mcpu=gfx900 -disassemble -show-encoding < %s 2>&1 | FileCheck -check-prefix=GCN %s
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx900 -disassemble -show-encoding < %s 2>&1 >/dev/null | FileCheck -check-prefix=GCN-ERR %s
```
Comment (arsenm) [inline llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:1]: Can you use -filetype=null instead of redirect stdout to /dev/null

Comment (kosarev) [reply llvm/test/MC/Disassembler/AMDGPU/decode-err.txt]: Done.

**Review by arsenm — APPROVED** [review]

File: llvm/test/MC/Disassembler/AMDGPU/decode-err.txt (hunk 1 above)
Comment (jayfoad) [reply llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:51]: Yes, I noticed that recently thanks to `filecheck_lint.py`. If you add a RUN line for gfx90a, the check fails. It needs more investigation.

File: llvm/test/MC/Disassembler/AMDGPU/decode-err.txt [hunk 3]
```diff
@@ -1,17 +1,18 @@
-# RUN: llvm-mc -triple=amdgcn -mcpu=gfx900 -disassemble -show-encoding < %s 2>&1 | FileCheck -check-prefix=GCN %s
-# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1100 -disassemble -show-encoding < %s 2>&1 | FileCheck -check-prefixes=GFX11,W32 %s
-# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1100 -mattr=-WavefrontSize32,+WavefrontSize64 -disassemble -show-encoding < %s 2>&1 | FileCheck -check-prefixes=GFX11,W64 %s
-# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -disassemble -show-encoding < %s 2>&1 | FileCheck -check-prefix=GFX12 %s
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx900 -disassemble -show-encoding -filetype=null < %s 2>&1 | FileCheck -check-prefix=GCN-ERR %s
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1100 -disassemble -show-encoding < %s 2>&1 | FileCheck -check-prefixes=W32 %s
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1100 -mattr=-WavefrontSize32,+WavefrontSize64 -disassemble -show-encoding < %s 2>&1 | FileCheck -check-prefixes=W64 %s
```
[the remark this answers is not in the document]
Comment (jayfoad) [reply llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:3]: > You also should never need to specify the negated form

@arsenm why do you say that? If you don't specify the negated form, you can end up with both wavefrontsize32 and wavefrontsize64 set, which is sometimes harmless but sometimes not. Related: #86957

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:3]: The subtarget features are processed as a state machine assigning values to the subtarget variable as each case is encountered. Code shouldn't be relying on a boolean is this feature present, but checking the resulting value of WavefrontSize

[the remark this answers is not in the document]
Comment (jayfoad) [reply llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:3]: Plenty of code goes straight to the feature bits:
```
$ grep -r getFeatureBits.*FeatureWavefrontSize lib/Target/AMDGPU/
lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:  if (isGFX10Plus() && getFeatureBits()[AMDGPU::FeatureWavefrontSize64] &&
lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:      !getFeatureBits()[AMDGPU::FeatureWavefrontSize32]) {
lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:      if (!getFeatureBits()[AMDGPU::FeatureWavefrontSize32])
lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:      if (!getFeatureBits()[AMDGPU::FeatureWavefrontSize64])
lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:      if (!getFeatureBits()[AMDGPU::FeatureWavefrontSize32])
lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:      if (!getFeatureBits()[AMDGPU::FeatureWavefrontSize64])
lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCKernelDescriptor.cpp:    if (STI->getFeatureBits().test(FeatureWavefrontSize32))
lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:  if (STI->getFeatureBits().test(FeatureWavefrontSize16))
lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:  if (STI->getFeatureBits().test(FeatureWavefrontSize32))
lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:      STI->getFeatureBits().test(FeatureWavefrontSize32);
lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:      STI->getFeatureBits().test(FeatureWavefrontSize32);
lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:  bool IsWave32 = STI->getFeatureBits().test(FeatureWavefrontSize32);
lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:  if (STI->getFeatureBits().test(FeatureWavefrontSize32)) {
```

[the remark this answers is not in the document]
Comment (kosarev) [reply llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:3]: > The capitalized wavefrontsize features shouldn't even parse?

It seems to get lowered before processing. Not sure if it should or shouldn't be done, though.

Re negations: that would be an unrelated change anyway, so I will submit this as is, if no objections.

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/MC/Disassembler/AMDGPU/decode-err.txt:3]: Plenty of code is broken?

**Pull request merged**